test/zdtm/static/sk-unix-dgram-ghost.c

```diff
@@ -0,0 +1,220 @@
+#ifndef _GNU_SOURCE
+# define _GNU_SOURCE
+#endif
+
+#include <stdlib.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <string.h>
+#include <stdbool.h>
+#include <limits.h>
+#include <errno.h>
+#include <dirent.h>
+#include <sys/wait.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <sys/stat.h>
+
+#include "zdtmtst.h"
+
+const char *test_doc	= "Check data of bound ghost DGRAM unix socket and possibility to connect";
+const char *test_author	= "Alexander Mikhalitsyn <alexander@mihalicyn.com>";
+
+/*
+ * PROCESS_NUM | FREEZE_FREQ
+ * 3           | 1 / 5
+ * 4           | 1 / 5
+ * 5           | 2 / 5
+ * 10          | 10 / 10
+ */
+#define PROCESSES_NUM 10
+
+#define MSG "hello"
+char filename[PATH_MAX];
+char *dirname;
+TEST_OPTION(dirname, string, "directory name", 1);
+
+static int fill_sock_name(struct sockaddr_un *name, const char *filename)
+{
+	char *cwd;
+
+	cwd = get_current_dir_name();
+	if (strlen(filename) + strlen(cwd) + 1 >= sizeof(name->sun_path)) {
+		pr_err("Name %s/%s is too long for socket\n",
+		       cwd, filename);
+		return -1;
+	}
+
+	name->sun_family = AF_LOCAL;
+	ssprintf(name->sun_path, "%s/%s", cwd, filename);
+	return 0;
+}
+
+static int client(int i, task_waiter_t t)
+{
+	struct sockaddr_un addr;
+	int sk;
+
+	sk = socket(AF_UNIX, SOCK_DGRAM, 0);
+	if (sk < 0) {
+		pr_perror("open client %d", i);
+		return 1;
+	}
+
+	if (fill_sock_name(&addr, filename) < 0) {
+		pr_err("%s is too long for socket\n", filename);
+		return 1;
+	}
+
+	if (connect(sk, (struct sockaddr *) &addr, sizeof(struct sockaddr_un)) < 0) {
+		pr_perror("connect failed %d", i);
+		return 1;
+	}
+
+	/* we are ready to c/r */
+	task_waiter_complete(&t, 1);
+
+	/* wait for server let us to send data */
+	task_waiter_wait4(&t, 2);
+
+	test_msg("child %d: lets send\n", i);
+
+	if (send(sk, MSG, sizeof(MSG), 0) != sizeof(MSG)) {
+		pr_perror("send failed %d", i);
+		return 1;
+	}
+
+	test_msg("child %d: MSG was sent\n", i);
+
+	/* notify server that we sent data */
+	task_waiter_complete(&t, 3);
+
+	return 0;
+}
+
+void child_exited(int signo)
+{
+	int status;
+
+	if (wait(&status) == -1) {
```

Comment:
you need to call wait in a loop because sigchld from a few children can be merged.

Reply:
Sure. Fixed!